src/transform/src/will_distinct.rs

```diff
@@ -0,0 +1,144 @@
+
+use mz_expr::{MirRelationExpr, MirScalarExpr};
+
+use crate::analysis::{DerivedView, NonNegative};
+
+use crate::{TransformCtx, TransformError};
+
+/// Pushes down the information that a collection will be subjected to a `Distinct` on specific columns.
+///
+/// This intends to recognize idiomatic stacks of `UNION` from SQL, which look like `Distinct` over `Union`, potentially
+/// over other `Distinct` expressions. It is only able to see patterns of `DISTINCT UNION DISTINCT, ..` and other operators
+/// in between will prevent the necessary insight to remove the second `DISTINCT`.
+///
+/// There are other potential optimizations, for example fusing multiple reads of the same source, where distinctness may
+/// mean they could be a single read (with additional requirements).
+#[derive(Debug, Default)]
+pub struct WillDistinct;
+
+impl crate::Transform for WillDistinct {
+    #[mz_ore::instrument(
+        target = "optimizer"
+        level = "trace",
+        fields(path.segment = "will_distinct")
+    )]
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        ctx: &mut TransformCtx,
+    ) -> Result<(), TransformError> {
+        // Perform bottom-up equivalence class analysis.
+        use crate::analysis::DerivedBuilder;
+        let mut builder = DerivedBuilder::new(ctx.features);
+        builder.require(NonNegative);
+        let derived = builder.visit(relation);
+        let derived = derived.as_view();
+
+        self.apply(relation, derived);
+
+        mz_repr::explain::trace_plan(&*relation);
+        Ok(())
+    }
+}
+
+impl WillDistinct {
+    fn apply(&self, expr: &mut MirRelationExpr, derived: DerivedView) {
+        // Record a list of expressions paired with optional "will distinct by" columns.
+        let mut todo = vec![(expr, derived, None::<&[MirScalarExpr]>)];
+        while let Some((expr, derived, distinct_by)) = todo.pop() {
+            // If we find a `Distinct` expression in the shadow of another `Distinct` that will apply to its key columns,
+            // we can remove this `Distinct` operator as the distinctness will be enforced by the other expression.
+            if let (
+                MirRelationExpr::Reduce {
+                    input,
+                    group_key,
+                    aggregates,
+                    ..
+                },
+                Some(columns),
+            ) = (&*expr, distinct_by)
+            {
+                if aggregates.is_empty()
+                    && columns
+                        .iter()
+                        .enumerate()
+                        .all(|(c, e)| e == &MirScalarExpr::Column(c))
+                    && columns.len() == group_key.len()
+                {
+                    let arity = input.arity();
+                    *expr = input
+                        .clone()
+                        .map(group_key.clone())
+                        .project((arity..arity + group_key.len()).collect::<Vec<_>>());
```

Comment:
This stops the traversal, by not pushing anything into `todo`. This would mean that processing a stack of `UNION`s would be quadratic, because of the fixpoint loop around `FuseAndCollapse` calling the new transform again and again. Would it be possible to continue the traversal here? Or what that be tricky due to `Analysis` results getting misaligned due to the tree changes?

Reply:
Thinking about this more, would it be ok to simply jump the traversal to `input` and jump to the `Analysis` result for `input` as well? They wouldn't get misaligned this way, right?

Reply:
Another thing: What do you think about putting this map+project into an MFP and calling `MFP::optimize` on it? This map+project would often optimize to a simple project.

I don't see anything near `WillDistinct` in the optimizer pipeline which would optimize this MFP (at least, not in `FuseAndCollapse`'s fixpoint loop), so if we don't optimize here, then we might carry around this map+project for a long time.

Reply:
Yeah, it's possible to keep going, just .. a bit horrid with our smart constructors. If you call `project(..).map(..)` .. maybe you have a `Project { Map { .. } }` but maybe not if the projection is the identity idk. There's probably a way to do it though; I'll keep thinking.

Reply:
Two changes, neither of which are the one you want:
1. Replaced `.clone()` with `.take_dangerous()`, after playing around with how things are borrowed.
2. Continuing the descent in the case that the `aggregates.is_empty()` test fails.

Reply:
Descent now continues with the non-smart constructors!